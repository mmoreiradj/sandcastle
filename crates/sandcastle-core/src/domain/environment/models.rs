mod command;
pub mod config;
mod environment;
mod gitops;
mod vcs;

pub use command::*;
pub use environment::*;
pub use vcs::*;
