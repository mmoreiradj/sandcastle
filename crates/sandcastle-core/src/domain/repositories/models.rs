use std::{backtrace::Backtrace, str::FromStr};

use octocrab::Octocrab;
use snafu::ResultExt;

use crate::error::{SandcastleError, ServiceErrorCode};

#[derive(Debug, Clone)]
pub enum GitOpsPlatformType {
    ArgoCD,
}

impl FromStr for GitOpsPlatformType {
    type Err = SandcastleError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(match s {
            "argocd" => GitOpsPlatformType::ArgoCD,
            _ => {
                return Err(SandcastleError::Service {
                    code: ServiceErrorCode::InvalidConfiguration,
                    message: "Invalid gitops platform".to_string(),
                    reason: s.to_string(),
                    backtrace: Backtrace::capture(),
                });
            }
        })
    }
}

#[derive(Debug, Clone)]
pub struct RepositoryConfiguration {
    pub repository_url: String,
    pub authentication: Authentication,
    pub gitops_platform: GitOpsPlatformType,
}

impl TryFrom<&RepositoryConfiguration> for Octocrab {
    type Error = SandcastleError;

    fn try_from(value: &RepositoryConfiguration) -> Result<Self, Self::Error> {
        match value.authentication.clone() {
            Authentication::GitHubApp(auth) => {
                let key = jsonwebtoken::EncodingKey::from_rsa_pem(auth.private_key.as_bytes())
                    .whatever_context(format!(
                        "Failed to encode private key for GitHub app {}",
                        auth.app_id
                    ))?;
                let octocrab = Octocrab::builder()
                    .app(auth.app_id.into(), key)
                    .build()
                    .map_err(|e| SandcastleError::Service {
                        code: ServiceErrorCode::GitHubAppAuthentication,
                        message: e.to_string(),
                        reason: auth.app_id.to_string(),
                        backtrace: Backtrace::capture(),
                    })?;
                let octocrab = octocrab
                    .installation(auth.installation_id.into())
                    .map_err(|e| SandcastleError::Service {
                        code: ServiceErrorCode::GitHubAppAuthentication,
                        message: e.to_string(),
                        reason: auth.app_id.to_string(),
                        backtrace: Backtrace::capture(),
                    })?;
                Ok(octocrab)
            }
        }
    }
}

#[derive(Debug, Clone)]
pub enum Authentication {
    GitHubApp(GitHubAppAuthentication),
}

#[derive(Debug, Clone)]
pub struct GitHubAppAuthentication {
    pub app_id: u64,
    pub installation_id: u64,
    pub private_key: String,
}
