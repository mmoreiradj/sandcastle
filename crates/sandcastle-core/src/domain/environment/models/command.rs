use std::backtrace::Backtrace;

use crate::error::{SandcastleError, ServiceErrorCode};

#[derive(Debug, Clone, PartialEq)]
pub enum ReconcileTrigger {
    CommentCommand(Command),
    PushEvent,
    PullRequestClosed,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Command {
    Deploy,
    Destroy,
}

impl Command {
    pub fn parse(comment_body: &str) -> Result<Option<Self>, SandcastleError> {
        let trimmed = comment_body.trim().to_lowercase();

        if trimmed.starts_with("sandcastle deploy") || trimmed == "sandcastle deploy" {
            Ok(Some(Command::Deploy))
        } else if trimmed.starts_with("sandcastle destroy") || trimmed == "sandcastle destroy" {
            Ok(Some(Command::Destroy))
        } else if trimmed.starts_with("sandcastle") {
            Err(SandcastleError::Service {
                code: ServiceErrorCode::InvalidConfiguration,
                message: format!("Unknown sandcastle command: {}", comment_body),
                reason: "Expected 'sandcastle deploy' or 'sandcastle destroy'".to_string(),
                backtrace: Backtrace::capture(),
            })
        } else {
            Ok(None)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_deploy() {
        let result = Command::parse("sandcastle deploy").unwrap();
        assert_eq!(result, Some(Command::Deploy));

        let result = Command::parse("  SANDCASTLE DEPLOY  ").unwrap();
        assert_eq!(result, Some(Command::Deploy));

        let result = Command::parse("sandcastle deploy with extra args").unwrap();
        assert_eq!(result, Some(Command::Deploy));
    }

    #[test]
    fn test_parse_destroy() {
        let result = Command::parse("sandcastle destroy").unwrap();
        assert_eq!(result, Some(Command::Destroy));

        let result = Command::parse("  SANDCASTLE DESTROY  ").unwrap();
        assert_eq!(result, Some(Command::Destroy));
    }

    #[test]
    fn test_parse_non_sandcastle_comment() {
        let result = Command::parse("This is just a regular comment").unwrap();
        assert_eq!(result, None);

        let result = Command::parse("LGTM!").unwrap();
        assert_eq!(result, None);
    }

    #[test]
    fn test_parse_invalid_sandcastle_command() {
        let result = Command::parse("sandcastle unknown");
        assert!(result.is_err());
    }
}
