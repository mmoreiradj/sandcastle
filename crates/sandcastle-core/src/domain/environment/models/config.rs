use std::{backtrace::Backtrace, ops::Deref, str::FromStr, sync::OnceLock};

use regex::Regex;
use serde::{Deserialize, Serialize};
use serde_yaml::Value;

use crate::error::{SandcastleError, ServiceErrorCode};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum BuiltinConfigKey {
    EnvironmentName,
    RepoURL,
    TargetRevision,
    LastCommitSHA,
    PRNumber,
}

impl BuiltinConfigKey {
    pub fn from_key(key: &str) -> Option<Self> {
        match key {
            ".Sandcastle.EnvironmentName" => Some(BuiltinConfigKey::EnvironmentName),
            ".Sandcastle.RepoURL" => Some(BuiltinConfigKey::RepoURL),
            ".Sandcastle.TargetRevision" => Some(BuiltinConfigKey::TargetRevision),
            ".Sandcastle.LastCommitSHA" => Some(BuiltinConfigKey::LastCommitSHA),
            ".Sandcastle.PRNumber" => Some(BuiltinConfigKey::PRNumber),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct ConfigPath(String);

impl Deref for ConfigPath {
    type Target = String;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

static VALIDATION_REGEX: OnceLock<Regex> = OnceLock::new();

fn get_regex() -> &'static Regex {
    VALIDATION_REGEX.get_or_init(|| {
        Regex::new(r#"^\.(?:Sandcastle|Custom)(?:\.[A-Za-z0-9]+)+$"#)
            .expect("Failed to compile regex")
    })
}

impl FromStr for ConfigPath {
    type Err = SandcastleError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if !get_regex().is_match(s) {
            return Err(SandcastleError::Service {
                code: ServiceErrorCode::InvalidConfiguration,
                message: "Invalid config path, must match .<Sandcastle|Custom>.<key>. Ex: .Sandcastle.EnvironmentName or .Custom.baseDomain".to_string(),
                reason: s.to_string(),
                backtrace: Backtrace::capture(),
            });
        }

        Ok(ConfigPath(s.to_string()))
    }
}

/// Represents the sandcastle configuration from the application file
/// found in the repository.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SandcastleConfiguration {
    pub custom: SandcastleCustomValues,
    pub template: String,
}

pub type SandcastleCustomValues = Value;

impl SandcastleConfiguration {
    pub fn from_string(string: &str) -> Result<Self, SandcastleError> {
        match string.trim().trim_start_matches("---").split_once("---") {
            Some((config, template)) => {
                let custom_config = serde_yaml::from_str::<Value>(config)
                    .map_err(|e| SandcastleError::Service {
                        code: ServiceErrorCode::InvalidConfiguration,
                        message: e.to_string(),
                        reason: config.to_string(),
                        backtrace: Backtrace::capture(),
                    })?
                    .get("custom")
                    .unwrap_or(&Value::Null)
                    .clone();
                Ok(Self {
                    custom: custom_config,
                    template: template.to_string(),
                })
            }
            None => Err(SandcastleError::Service {
                code: ServiceErrorCode::InvalidConfiguration,
                message: "No configuration found in file".to_string(),
                reason: string.to_string(),
                backtrace: Backtrace::capture(),
            }),
        }
    }

    pub fn get_custom_value(&self, path: &str) -> Option<String> {
        let path_parts = path
            .trim_start_matches(".Custom.")
            .split(".")
            .map(|s| s.to_string())
            .collect::<Vec<String>>();
        let mut current = &self.custom;

        for part in path_parts {
            println!("Part: {}", part);
            current = current.get(part.as_str())?;
        }

        current.as_str().map(|s| s.to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_get_custom_value() {
        let custom = r#"
        custom:
          baseDomain: sandcastle.dev
          whatever:
            key: value
        ---
        whatever:
          key: value
        "#;
        let config = SandcastleConfiguration::from_string(custom).unwrap();
        let value = config.get_custom_value(".Custom.baseDomain");
        assert_eq!(value, Some("sandcastle.dev".to_string()));

        let value = config.get_custom_value(".Custom.baseDomain.subDomain");
        assert_eq!(value, None);

        let value = config.get_custom_value(".Custom.baseDomain.subDomain.subDomain");
        assert_eq!(value, None);

        let value = config.get_custom_value(".Custom.whatever.key");
        assert_eq!(value, Some("value".to_string()));
    }

    #[test]
    fn test_from_string() {
        let application_yaml =
            include_str!("../../../../tests/fixtures/example_application_1.yaml");
        let config = SandcastleConfiguration::from_string(application_yaml).unwrap();
        assert_eq!(
            config.get_custom_value(".Custom.baseDomain"),
            Some("sandcastle.dev".to_string())
        );
        assert_eq!(
            config.get_custom_value(".Custom.whatever.key"),
            Some("value".to_string())
        );
    }
}
