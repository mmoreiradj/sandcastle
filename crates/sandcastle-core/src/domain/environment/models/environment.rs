use std::backtrace::Backtrace;

use crate::{
    Result,
    domain::environment::models::{DownloadFileRequest, FetchPRLastCommitSHARequest},
    error::ServiceErrorCode,
};
use crate::{
    domain::environment::{
        models::config::{BuiltinConfigKey, SandcastleConfiguration},
        ports::{Reconcile, VCSService},
        services::{GitOpsPlatform, Vcs},
    },
    error::SandcastleError,
};
use octocrab::models::{
    Repository,
    webhook_events::{WebhookEvent, WebhookEventPayload},
};
use regex::Regex;

#[derive(Clone)]
pub struct ReconcileContext {
    /// The ID of the reconcile context
    pub id: String,
    /// The VCS context
    pub vcs: VcsContext,
    /// The VCS service
    pub vcs_service: Vcs,
    /// The GitOps platform service
    pub gitops_platform_service: GitOpsPlatform,
    /// Sandcastle configuration
    pub config: SandcastleConfiguration,
    /// The trigger that initiated this reconciliation
    pub trigger: super::ReconcileTrigger,
}

impl ReconcileContext {
    pub fn template(&self, template: &str) -> Result<String, SandcastleError> {
        let mut result = template.to_string();
        let r = Regex::new(r#"\{\{ (.*?) \}\}"#).unwrap();
        let replacements: Vec<(String, String)> = r
            .captures_iter(&result)
            .map(|capture| -> Result<(String, String), SandcastleError> {
                let full_match = capture.get(0).unwrap().as_str().to_string();
                let path = capture.get(1).unwrap().as_str().trim();
                let value =
                    self.get_config_value(path)
                        .ok_or_else(|| SandcastleError::Service {
                            code: ServiceErrorCode::InvalidConfiguration,
                            message: format!("Value not found for path: {}", path),
                            reason: path.to_string(),
                            backtrace: Backtrace::capture(),
                        })?;
                Ok((full_match, value))
            })
            .collect::<Result<Vec<_>, _>>()?;

        for (pattern, replacement) in replacements {
            result = result.replace(&pattern, &replacement);
        }
        Ok(result)
    }

    pub async fn from_github_event(
        id: String,
        event: WebhookEvent,
        payload: WebhookEventPayload,
        vcs_service: Vcs,
        gitops_platform_service: GitOpsPlatform,
    ) -> Result<Option<Self>> {
        match payload {
            WebhookEventPayload::IssueComment(payload) => {
                let comment_body = if let Some(body) = payload.comment.body {
                    body
                } else {
                    return Ok(None);
                };

                let command = super::Command::parse(&comment_body)?;
                let trigger = match command {
                    Some(cmd) => super::ReconcileTrigger::CommentCommand(cmd),
                    None => return Ok(None),
                };

                let repository = event.repository.unwrap();

                let last_commit_sha = vcs_service
                    .fetch_pr_last_commit_sha(FetchPRLastCommitSHARequest {
                        repository_id: (*repository.id),
                        pr_number: payload.issue.number,
                    })
                    .await?;

                let config =
                    Self::fetch_config(&vcs_service, &repository, &last_commit_sha).await?;

                Ok(Some(Self::build(
                    id,
                    VcsContext {
                        comment: CommentContext { body: comment_body },
                        repository: RepositoryContext::from(&repository),
                        pull_request: PullRequestContext {
                            number: payload.issue.number,
                            title: payload.issue.title.clone(),
                            last_commit_sha,
                        },
                    },
                    vcs_service,
                    gitops_platform_service,
                    config,
                    trigger,
                )))
            }
            WebhookEventPayload::Push(payload) => {
                let repository = event.repository.unwrap();
                let commit_sha = payload.after;

                let config = Self::fetch_config(&vcs_service, &repository, &commit_sha).await?;

                Ok(Some(Self::build(
                    id,
                    VcsContext {
                        comment: CommentContext {
                            body: String::new(),
                        },
                        repository: RepositoryContext::from(&repository),
                        pull_request: PullRequestContext {
                            number: 0,
                            title: format!("Push to {}", payload.r#ref),
                            last_commit_sha: commit_sha,
                        },
                    },
                    vcs_service,
                    gitops_platform_service,
                    config,
                    super::ReconcileTrigger::PushEvent,
                )))
            }
            WebhookEventPayload::PullRequest(payload) => {
                if payload.action != octocrab::models::webhook_events::payload::PullRequestWebhookEventAction::Closed {
                    return Ok(None);
                }

                let repository = event.repository.unwrap();
                let pr = payload.pull_request;
                let last_commit_sha = pr.head.sha;

                let config =
                    Self::fetch_config(&vcs_service, &repository, &last_commit_sha).await?;

                Ok(Some(Self::build(
                    id,
                    VcsContext {
                        comment: CommentContext {
                            body: String::new(),
                        },
                        repository: RepositoryContext::from(&repository),
                        pull_request: PullRequestContext {
                            number: pr.number,
                            title: pr.title.unwrap_or_default(),
                            last_commit_sha,
                        },
                    },
                    vcs_service,
                    gitops_platform_service,
                    config,
                    super::ReconcileTrigger::PullRequestClosed,
                )))
            }
            _ => Ok(None),
        }
    }

    async fn fetch_config(
        vcs_service: &Vcs,
        repository: &Repository,
        commit_sha: &str,
    ) -> Result<SandcastleConfiguration> {
        let configuration_file_content = vcs_service
            .download_file(DownloadFileRequest {
                repository_id: (*repository.id),
                path: ".github/sandcastle.yaml".to_string(),
                r#ref: commit_sha.to_string(),
                content_type: "application/yaml".to_string(),
            })
            .await?;

        SandcastleConfiguration::from_string(&configuration_file_content)
    }

    fn build(
        id: String,
        vcs: VcsContext,
        vcs_service: Vcs,
        gitops_platform_service: GitOpsPlatform,
        config: SandcastleConfiguration,
        trigger: super::ReconcileTrigger,
    ) -> Self {
        Self {
            id,
            vcs,
            vcs_service,
            gitops_platform_service,
            config,
            trigger,
        }
    }

    fn get_config_value(&self, path: &str) -> Option<String> {
        if path.starts_with(".Sandcastle.") {
            self.get_builtin_config_value(path)
        } else {
            self.config.get_custom_value(path)
        }
    }

    fn get_builtin_config_value(&self, key: &str) -> Option<String> {
        let key = BuiltinConfigKey::from_key(key)?;
        match key {
            BuiltinConfigKey::EnvironmentName => Some(self.vcs.repository.name.clone()),
            BuiltinConfigKey::RepoURL => Some(self.vcs.repository.url.clone()),
            BuiltinConfigKey::TargetRevision => Some(self.vcs.pull_request.last_commit_sha.clone()),
            BuiltinConfigKey::LastCommitSHA => Some(self.vcs.pull_request.last_commit_sha.clone()),
            BuiltinConfigKey::PRNumber => Some(self.vcs.pull_request.number.to_string()),
        }
    }

    pub fn labels(&self) -> Vec<(String, String)> {
        vec![
            (
                "environment".to_string(),
                self.config
                    .get_custom_value(".Sandcastle.EnvironmentName")
                    .unwrap_or_default(),
            ),
            ("repository".to_string(), self.vcs.repository.name.clone()),
            (
                "pull-request".to_string(),
                self.vcs.pull_request.number.to_string(),
            ),
        ]
    }
}

#[derive(Debug, Clone)]
pub struct VcsContext {
    pub repository: RepositoryContext,
    pub pull_request: PullRequestContext,
    pub comment: CommentContext,
}

#[derive(Debug, Clone)]
pub struct RepositoryContext {
    /// The name of the repository
    pub name: String,
    /// Whether the repository is private
    pub private: bool,
    /// The base URI of the repository
    pub url: String,
}

impl From<&Repository> for RepositoryContext {
    fn from(value: &Repository) -> Self {
        Self {
            name: value.name.clone(),
            private: value.private.unwrap_or(false),
            // this is always Some
            url: value.html_url.as_ref().unwrap().to_string(),
        }
    }
}

#[derive(Debug, Clone)]
pub struct PullRequestContext {
    pub number: u64,
    pub title: String,
    pub last_commit_sha: String,
}

#[derive(Debug, Clone)]
pub struct CommentContext {
    pub body: String,
}

/// Action to create or update a GitOps Application
#[derive(Debug, Clone)]
pub struct CreateOrUpdateArgocdApplicationAction {
    /// The GitOps File
    pub applications: Vec<String>,
}

/// Action to delete an Argocd Application
#[derive(Debug, Clone)]
pub struct DeleteArgocdApplicationAction {
    /// The GitOps File
    pub applications: Vec<String>,
}

#[derive(Debug, Clone)]
pub enum ReconcileActions {
    CreateOrUpdateArgocdApplication(CreateOrUpdateArgocdApplicationAction),
    DeleteArgocdApplication(DeleteArgocdApplicationAction),
}

impl ReconcileActions {
    pub async fn reconcile(&self, context: ReconcileContext) -> Result<(), SandcastleError> {
        match self {
            ReconcileActions::CreateOrUpdateArgocdApplication(action) => {
                action.reconcile(context).await
            }
            ReconcileActions::DeleteArgocdApplication(action) => action.reconcile(context).await,
        }
    }
}

#[cfg(test)]
mod tests {
    use googletest::prelude::*;
    use octocrab::models::webhook_events::WebhookEventType;

    use crate::domain::environment::ports::{MockGitOpsPlatformService, MockVCSService};
    use googletest::Result;

    use super::*;

    async fn test_context() -> ReconcileContext {
        let config = SandcastleConfiguration::from_string(include_str!(
            "../../../../tests/fixtures/example_application_1.yaml"
        ))
        .unwrap();
        let context = ReconcileContext {
            id: "1".to_string(),
            vcs: VcsContext {
                repository: RepositoryContext {
                    name: "sandcastle-monorepo-test".to_string(),
                    private: false,
                    url: "https://github.com/mmoreiradj/sandcastle-monorepo-test".to_string(),
                },
                pull_request: PullRequestContext {
                    number: 1,
                    title: "feat: make unreasonable promises".to_string(),
                    last_commit_sha: "1234567890".to_string(),
                },
                comment: CommentContext {
                    body: "sandcastle deploy".to_string(),
                },
            },
            vcs_service: Vcs::MockVCS(MockVCSService::new()),
            gitops_platform_service: GitOpsPlatform::MockGitOpsPlatform(
                MockGitOpsPlatformService::new(),
            ),
            config: config,
            trigger: crate::domain::environment::models::ReconcileTrigger::CommentCommand(
                crate::domain::environment::models::Command::Deploy,
            ),
        };
        context
    }

    #[tokio::test]
    async fn test_small_template() {
        let template = "{{ .Sandcastle.EnvironmentName }}";
        let context = test_context().await;
        let result = context.template(template).unwrap();
        assert_eq!(result, "sandcastle-monorepo-test");
    }

    #[tokio::test]
    async fn test_large_template() {
        let template = include_str!("../../../../tests/fixtures/example_application_1.yaml");
        let context = test_context().await;
        let result = context.template(template).unwrap();
        insta::assert_snapshot!(result);
    }

    #[tokio::test]
    #[gtest]
    async fn test_from_github_event_issue_comment() -> Result<()> {
        let comment_payload_str =
            include_str!("../../../../tests/fixtures/github/issue_comment_webhook_event.json");
        let comment_payload =
            serde_json::from_str::<serde_json::Value>(comment_payload_str).unwrap();
        let event =
            WebhookEvent::try_from_header_and_body("issue_comment", comment_payload_str).unwrap();
        let webhook_event_type = WebhookEventType::IssueComment;
        let payload = webhook_event_type
            .parse_specific_payload(comment_payload)
            .unwrap();

        let mut mock_vcs_service = MockVCSService::new();
        mock_vcs_service
            .expect_fetch_pr_last_commit_sha()
            .returning(move |_| Ok("1234567890".to_string()));
        mock_vcs_service.expect_download_file().returning(move |_| {
            let template = include_str!("../../../../tests/fixtures/example_application_1.yaml");
            Ok(template.to_string())
        });
        let context = ReconcileContext::from_github_event(
            "1".to_string(),
            event,
            payload,
            Vcs::MockVCS(mock_vcs_service),
            GitOpsPlatform::MockGitOpsPlatform(MockGitOpsPlatformService::new()),
        )
        .await
        .unwrap()
        .unwrap();
        expect_that!(
            context.vcs.repository,
            matches_pattern!(RepositoryContext {
                name: eq("sandcastle-monorepo-test"),
                private: eq(&true),
                url: eq("https://github.com/mmoreiradj/sandcastle-monorepo-test"),
            })
        );
        expect_that!(
            context.vcs.pull_request,
            matches_pattern!(PullRequestContext {
                number: eq(&1),
                title: eq("feat: make unreasonable promises"),
                last_commit_sha: eq("1234567890"),
            })
        );
        expect_that!(
            context.vcs.comment,
            matches_pattern!(CommentContext {
                body: eq("sandcastle deploy"),
            })
        );
        Ok(())
    }
}
