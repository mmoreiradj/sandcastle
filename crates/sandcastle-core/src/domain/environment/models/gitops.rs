#[derive(Debug, Clone)]
pub struct CreateOrUpdateArgocdApplicationRequest {
    pub applications: Vec<String>,
    pub labels: Vec<(String, String)>,
}
