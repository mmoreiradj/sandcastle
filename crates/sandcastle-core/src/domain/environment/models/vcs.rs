#[derive(Debug, Clone)]
pub struct DownloadFileRequest {
    pub uri: String,
    pub content_type: String,
}
