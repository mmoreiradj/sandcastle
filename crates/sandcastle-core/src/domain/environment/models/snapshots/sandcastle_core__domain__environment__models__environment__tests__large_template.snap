---
source: crates/sandcastle-core/src/domain/environment/models/environment.rs
expression: result
---
---
custom:
  baseDomain: "sandcastle.dev"
  whatever:
    key: "value"
global:
  images:
    frontend:
      # Image hint to find the images in:
      # - github checks
      # - container registry
      hint: "ghcr.io/mmoreiradj/sandcastle-monorepo-test/frontend"
---
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: frontend-sandcastle-monorepo-test-1
  namespace: argocd
spec:
  project: default
  source:
    path: "charts/frontend" 
    repoURL: "https://github.com/mmoreiradj/sandcastle-monorepo-test"
    targetRevision: "1234567890"
    helm:
      valuesObject:
        ingress:
          enabled: true
          className: "traefik"
          hosts:
            - host: "frontend-sandcastle-monorepo-test-1.sandcastle.dev"
              paths:
                - path: /
                  pathType: ImplementationSpecific
        image:
          tag: "1234567890"
  destination:
    server: "https://kubernetes.default.svc"
    namespace: "sandcastle-monorepo-test-1"
